### ISSUE TITLE
"YADIF Bob" deinterlacing setting is ignored

### ISSUE THREAD

**Issue by DragonQ** [jellyfin/jellyfin#4314]
In Jellyfin server -> Settings -> Playback -> Transcoding, I've set the deinterlacing method to "YADIF Bob" (which is referred to everywhere else on the planet as YADIF 2x but I digress...). However, when I use the Jellyfin app on my phone to play 50i media, they are played back at 25p. The log contains this line, showing that YADIF 2x is **not** being used:

`[2020-10-12 13:27:32.683 +00:00] [INF] [132] MediaBrowser.Api.Playback.Hls.DynamicHlsService: /usr/lib/jellyfin-ffmpeg/ffmpeg -ss 00:04:39.000 -noaccurate_seek -f matroska,webm -i file:"/media/TV Programmes/8 Out of 10 Cats Does Countdown/8 Out of 10 Cats Does Countdown - Series 1/8 Out of 10 Cats Does Countdown S01E01 [1080i Upscaled].mkv" -map_metadata -1 -map_chapters -1 -threads 8 -map 0:0 -map 0:1 -map -0:s -codec:v:0 libx264 -pix_fmt yuv420p -preset fast -crf 22 -maxrate 8986510 -bufsize 17973020 -profile:v high -level 4.1 -x264opts:0 subme=0:me_range=4:rc_lookahead=10:me=dia:no_chroma_me:8x8dct=0:partitions=none  -force_key_frames:0 "expr:gte(t,279+n_forced*3)" -g 150 -keyint_min 150 -sc_threshold 0 -vf "yadif=0:-1:0,scale=trunc(min(max(iw\,ih*dar)\,1920)/2)*2:trunc(ow/dar/2)*2" -start_at_zero -vsync -1 -codec:a:0 libmp3lame -ac 2 -ab 384000  -copyts -avoid_negative_ts disabled -f hls -max_delay 5000000 -hls_time 3 -individual_header_trailer 0 -hls_segment_type mpegts -start_number 93 -hls_segment_filename "/config/transcodes/9f7ef09e6b18b78f7988b950c690361d%d.ts" -hls_playlist_type vod -hls_list_size 0 -y "/config/transcodes/9f7ef09e6b18b78f7988b950c690361d.m3u8"`

Note the `yadif=0:-1:0` part. For YADIF 2x/bob, this should be `yadif=1:-1:0`. I've attached debug logs but they don't seem to show anything useful.
[jellyfin.zip](https://github.com/jellyfin/jellyfin/files/5365458/jellyfin.zip)

**Comment by telans** [comment]
Looks like this may be intended behaviour

https://github.com/jellyfin/jellyfin/blob/49ac4c4044b1777dc3a25544aead7b0b15b953e8/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L2104-L2105

**Comment by DragonQ** [comment]
It can't be intended behaviour for 50i files to not be deinterlaced to 50p. Since every other device and media renderer I use correctly identifies the videos as 50i and deinterlaces them to 50p, there's no reason Jellyfin should refuse to do this.

That line of code suggests that for these videos, `videoStream?.RealFrameRate` has a value of `null` or `50`, leading to `doubleRateDeinterlace` being set to false. I am not sure what `videoStream?.RealFrameRate` represents but if this line of code's logic is correct, it presumably should be `25` for 50i content. Either that or the logic is incorrect.

Here's a 10s sample of the file I was trying to play: [8 Out of 10 Cats Does Countdown S05E01 [1080i]-010.zip](https://github.com/jellyfin/jellyfin/files/5400498/8.Out.of.10.Cats.Does.Countdown.S05E01.1080i.-010.zip).

**Comment by telans** [comment]
`videoStream?.RealFrameRate` looks like it goes back to `r_frame_rate` which is `50/1` for the file you provided.

```
ffprobe -select_streams v:0 -show_entries stream=r_frame_rate 8\ Out\ of\ 10\ Cats\ Does\ Countdown\ S05E01\ \[1080i\]-010.mkv
```

**Comment by DragonQ** [comment]
I see. I think it's very unlikely that the file container's information is wrong or non-standard, given it was created using an official application with default parameters. So the question is why does Jellyfin behave in this way and what is the best way to change it so that the files are deinterlaced as specified by the user?

I've never heard of 120i video, so I'm assuming the purpose of that guard in the code is to prevent 60i files that are _incorrectly labelled_ being deinterlaced to 120p (or 50i to 100p). If so, is the guard needed at all? Does YADIF 2x really struggle with such mislabelled files? It'd be interesting to know what files this code was tested against? If the guard really is needed for some strange edge cases then it should be altered to be less bludgeoning - after all, workarounds for dodgy files should not negatively affect correctly formatted files. Checking the scan type and original frame rate would be a reasonable starting point, although there may be a simpler way.

**Comment by telans** [comment]
@orryverducci added the check in https://github.com/jellyfin/jellyfin/pull/3877, so perhaps they might have some insights about this

**Comment by orryverducci** [comment]
I borrowed that check from the existing code I replaced, so I didn't thoroughly test it as it was existing behaviour that I assumed was there for good reason.

My understanding/purpose of the check is exactly as @DragonQ has suggested, it's to guard against mislabelled or 'non standard' files so that it doesn't try 2x deinterlacing to anything higher than 60fps. That check isn't exclusive to YADIF, it also applies to BWDIF and the hardware deinterlacers which may not cope (in real time at least) with deinterlacing high frame rates, so it seems like a reasonable guard to have in place. I'm not aware of any valid interlaced files that are higher than 60i.

In the case of this file I would have expected `videoStream?.RealFrameRate` to be 25, as it's 25 frames per second, but 50 fields. If FFprobe's reporting it as 50fps, that is different to what I expected it to report and would definitely fail the check, so I'll have to take a second look at this as that's definitely a bug.

I should also add I don't think my changes have been backported to the current 10.6 release, which I assume @DragonQ is using as the changes remove 'YADIF Bob' and replace it with a double rate check box. If that's the case, that line of code isn't in it so it's not the problem here. But as I said I borrowed the check from the existing code, so it's probably is that check but located elsewhere.

**Comment by DragonQ** [comment]
@orryverducci, I am indeed using v10.6.4. It may be that different containers report different values for "frame rate", I'm not sure. It looks like you re-used the same guard in a different part of the code in your PR, so this bug may still affect the next release.

**Comment by orryverducci** [comment]
> @orryverducci, I am indeed using v10.6.4. It may be that different containers report different values for "frame rate", I'm not sure. It looks like you re-used the same guard in a different part of the code in your PR, so this bug may still affect the next release.

Yes it probably does. My quick tests were with TS and MP4 files, so it's very possible the MKV container handles it different and I missed it.

I've got your file to test it with so I'll look in to it. Conveniently I'm in the UK so I'll make some recordings from Channel 4 HD and the other channels in different formats to check they all work.

**Comment by orryverducci** [comment]
I've done some investigation and it turns out the problem is due to the Matroska container. MKV's don't store a frame rate like the MPEG containers do, so instead it is calculated from the display duration for each block in the file.

As the blocks usually contain frames this is normally OK and results in the right frame rate being displayed. However for interlaced files these blocks contains fields instead of frames, and so the frame rate is incorrectly calculated and displayed as being double what it should be. This problem affects both FFprobe and MediaInfo.

In this case it's resulted in FFprobe (and therefore Jellyfin) thinking this file is 50fps instead of 25fps, and so double rate (bob) deinterlacing is being disabled.

I've submitted a pull request (#4369) which fixes the probe so the right frame rate is stored, which also fixes this issue.

**Comment by DragonQ** [comment]
Any updates on this? I know the revert commit mentions that the fix might have to be in ffmpeg but many media players can correctly figure out whether deinterlacing needs to be enabled or not so this doesn't feel like a bug that goes beyond Jellyfin.

**Comment by orryverducci** [comment]
Sorry for the lack of updates, I've been scratching my head over the best way to resolve this.

The problem is definitely in FFmpeg, not Jellyfin. As we know FFmpeg reports the file as being 50fps (50p not 50i), and if you try to use FFmpeg to transcode the file it will automatically try to create a 50fps encode even though the video is actually 25fps. If you try to deinterlace with FFmpeg it will incorrectly deinterlace to 100fps, based on its belief that the file is 50fps. As Jellyfin doesn't do any decoding/encoding itself and relies on FFmpeg, Jellyfin is affected.

This isn't unique to FFmpeg, VLC and MPV also report the wrong frame rate for this file (VLC gets away with it as it bobs the fields). As does MediaInfo, although as we know MediaInfo is clever enough to read the H.264 bitstream and return a second correct value.

I had hoped to work around it in Jellyfin by correcting FFmpeg's output and forcing it to 25/30fps, but as I've discovered not all interlaced MKV files report the wrong frame rate, depending on which H.264 encode settings have been used (PAFF or MBAFF). Basically files encoded by x264 work as intended, and recordings from broadcast TV vary (e.g. BBC One is ok, Channel 4 isn't). To work around it in Jellyfin we need to know if the blocks in the MKV contain frames or fields so we can make the appropriate correction, and FFmpeg doesn't tell us this, so unless anyone else has a good idea I think we have no choice but to throw it to the FFmpeg folks.

Ultimately MKV isn't a great container for interlaced content. It's lack of concept of frame rate and ambiguity in its specification (it doesn't say if blocks should contain fields or frames) is what's created this situation. For now my best advice would be to rewrap these files as MP4's or TS files.

I'm not giving up on this yet, just right now I'm short of a good solution.

**Comment by DragonQ** [comment]
Hmm that's a bit rubbish. If I open a random selection of my recordings from Channel 4 HD (DVB-S2) in MediaInfo, some of them say 25 FPS, some say 50 FPS. No consistency that I can work out.

**Comment by orryverducci** [comment]
> Hmm that's a bit rubbish. If I open a random selection of my recordings from Channel 4 HD (DVB-S2) in MediaInfo, some of them say 25 FPS, some say 50 FPS. No consistency that I can work out.

H.264 has a few different ways of encoding interlaced content, and different encoders and settings will choose different options. Broadcast encoders are usually configured to adapt to the content coming in, so depending on how much motion the incoming video has the encoders will often change their mind about which options to use, hence the lack of consistency.

When using containers that tell you the frame rate, such as the native TS stream, we don't have to worry about any of this. With MKV its much more complicated...

**Comment by DragonQ** [comment]
I just made some observations regarding the MKVs I mentioned above:
50i MKVs that are detected by MediaInfo as "50 FPS":
- do not deinterlace properly in Jellyfin
- have an "Original Frame Rate" field with a value of 25 FPS
50i MKVs that are detected by MediaInfo as "25 FPS":
- do deinterlace properly in Jellyfin
- do not have an "Original Frame Rate" field

Is it worth checking how MediaInfo gets the value for the "Original Frame Rate" field? The MKVToolNix wiki suggests that MediaInfo is looking directly at the underlying bitstream (in this case H.264) but I am not sure if MediaInfo uses ffprobe or some custom method for this.

**Comment by DragonQ** [comment]
Yes this is still an issue. It should be decided if it will actually be addressed or chalked up as "MKV and/or ffmpeg don't support interlaced video well and we aren't going to work around it".

**Comment by DragonQ** [comment]
3 years on, there must be a way to fix this issue. Interlaced content will be around forever, given it's still broadcast on HDTV to this day in many countries. This is still the primary (maybe _only_ these days) reason I don't use Jellyfin.

Even if it's an issue with the Matroska format or its tools, I refuse to believe it isn't possible to correctly deinterlace these files when other media players (including hardware) do this successfully with the same files.

**Comment by felix920506** [comment]
Tested on 10.10.7, still happens
